File: Python_TRR/7.py
# Viết bằng Python
# Bài 7: Số cách xếp n đồ vật vào k hộp.

from math import comb # Hàm comb(n, k) tính tổ hợp chập k của n

# Có thứ tự
def distribute_with_order(n, k): # n: số đồ vật, k: số hộp 
    return k**n

# Không có thứ tự
def distribute_without_order(n, k): # n: số đồ vật, k: số hộp 
    return comb(n + k - 1, k - 1) # Tổ hợp chập k-1 của n+k-1

# Ví dụ
n, k = 5, 3
print(f"Có thứ tự: {distribute_with_order(n, k)}") # 3^5 = 243
print(f"Không có thứ tự: {distribute_without_order(n, k)}") # C(5+3-1, 3-1) = C(7, 2) = 21


# Kết quả
# Có thứ tự: 243
# Không có thứ tự: 21

File: C_TRR/7.c
// Viết bằng C
// Bài 7: Số cách xếp n đồ vật vào k hộp.

#include <stdio.h>
#include <math.h>

long long comb(int n, int k)  // Hàm tính toán tổ hợp
{
    long long res = 1; // Khởi tạo biến res = 1
    for (int i = 1; i <= k; i++) // Vòng lặp từ i = 1 đến k
    {
        res *= n--; // res = res * n, n = n - 1
        res /= i; // res = res / i 
    }
    return res;
}

long long distribute_with_order(int n, int k)  
// Hàm tính số cách xếp có thứ tự
{
    return pow(k, n); // Trả về k^n 
}

long long distribute_without_order(int n, int k) 
// Hàm tính số cách xếp không có thứ tự
{
    return comb(n + k - 1, k - 1); 
    // Trả về tổ hợp (n + k - 1, k - 1)
}

int main() 
{
    int n = 5, k = 3; // Khai báo n = 5, k = 3 
    printf("Có thứ tự: %lld\n", distribute_with_order(n, k)); // In ra số cách xếp có thứ tự
    printf("Không có thứ tự: %lld\n", distribute_without_order(n, k)); // In ra số cách xếp không có thứ tự
    return 0;
}

// Kết quả:
// Có thứ tự: 243
// Không có thứ tự: 21
